### ISSUE TITLE
allow client plugins to depend on (and use) one another

### ISSUE THREAD

**Issue by andyroyle** [opentable/oc#128]
A-la hapi's `server.dependency(...)`.

**Comment by matteofigus** [comment]
:+1:

**Comment by andyroyle** [comment]
@matteofigus got an implementation for this working, wanted to get your opinion on how to pass the dependencies into the plugins (for the plugins to use).

Registering a dep is done like this `module.exports.dependencies = ['myplugin1', 'myplugin2'];`

So that the plugin can use it's dependencies inside the register function, we need to pass them in. My gut feeling is that we should change the signature for the register function to this:

`module.exports.register = function(options, deps, next){`

where `deps` is an object the similar to `context.plugins` inside server.js. 

Unfortunately this is a breaking change.

We could work around it by setting a property on the options object if you prefer.

**Comment by matteofigus** [comment]
All good for me. If we make deps optional not necessarily it is a breaking change right?
```js
...
register = function(options, deps, next){
if(typeof(deps) === 'function'){
  next = deps;
  deps = [];
}
...
```

**Comment by andyroyle** [comment]
the problem is that the register function is in client-code, so unless they've got that snippet in all their plugins then it'll break.

unless we try and wrap the register function and do some magic. But that might be a bit nuts

**Comment by matteofigus** [comment]
Yes, but we have semver for that (so if you implement plugins on top of a oc-registry you are safe and that shouldn't change anything on components) + I wrote this sentence on the very top of the README: `Disclaimer: This project is still under heavy development and the API is likely to change at any time. In case you would find any issues, check the troubleshooting page.` :dancers:

**Comment by matteofigus** [comment]
PS - I am trying to push to get a v1 out. In this perspective, I am trying to get the minimal requirements to start doing semver properly and then push majors for changes like this. If you wish to share your voice: https://github.com/opentable/oc/milestones/v1.0.0

**Comment by andyroyle** [comment]
:+1:

**Comment by andyroyle** [comment]
published in 0.20
